securedrop_client/gui/widgets.py

```diff
@@ -910,50 +907,76 @@ def setup(self, controller):
-        """
-        Update the list with the passed in list of sources.
-        """
-        # Delete widgets that no longer exist in source list
-        source_uuids = [source.uuid for source in sources]
-        deleted_uuids = []
-        for i in range(self.count()):
-            list_item = self.item(i)
-            list_widget = self.itemWidget(list_item)
-
-            if list_widget and list_widget.source_uuid not in source_uuids:
-                if list_item.isSelected():
-                    self.setCurrentItem(None)
-                del self.source_widgets[list_widget.source_uuid]
-                deleted_uuids.append(list_widget.source_uuid)
-                self.takeItem(i)
-                list_widget.deleteLater()
-
-        # Create new widgets for new sources
-        widget_uuids = [self.itemWidget(self.item(i)).source_uuid for i in range(self.count())]
-        for source in sources:
-            if source.uuid in widget_uuids:
-                try:
-                    self.source_widgets[source.uuid].update()
-                except sqlalchemy.exc.InvalidRequestError as e:
-                    logger.error(
-                        "Could not update SourceWidget for source %s; deleting it. Error was: %s",
-                        source.uuid,
-                        e
-                    )
-                    deleted_uuids.append(source.uuid)
-                    self.source_widgets[source.uuid].deleteLater()
-                    del self.source_widgets[list_widget.source_uuid]
-            else:
+    def update(self, sources: List[Source]):
+        """
+        Reset and update the list with the passed in list of sources.
+        """
+        # A flag to show if a source is currently selected (the current source
+        # doesn't have to be selected in a QListWidget -- it appears to default
+        # to whatever is in row 0, whether it's selected or not).
+        has_source_selected = False
+        if self.currentRow() > -1:
+            has_source_selected = self.item(self.currentRow()).isSelected()
+        current_source = self.get_current_source()
+        current_source_id = current_source and current_source.id
+        # Make a copy of the source_widgets that are currently displayed
+        # so that we can compare which widgets were removed. This means that
+        # the source was deleted, and we'll return this so we can also
+        # delete the corresponding SourceConversationWrapper.
+        existing_source_widgets = self.source_widgets.copy()
+
+        # When we call clear() to delete all SourceWidgets, we should
+        # also clear the source_widgets dict.
+        self.clear()
+        self.source_widgets = {}
+        sources.reverse()
+        self.add_source(existing_source_widgets, sources, current_source_id,
+                        has_source_selected)
+
+    def add_source(self, existing_source_widgets, sources, current_source_id,
```

Comment:
I think we should keep this method and use it with some modifications. I see how this improves performance when there are a large number of sources to add at once, but in order to also support not recreating all source widgets every time we sync (which is a huge performance hit for every other situation, especially when there are many existing source widgets) I think you can just update the signature to something like:

```
def add_source(self, sources_to_add, slice_size=1)
```

There's no reason to keep track of the `current_source_id` or `has_source_selected` or `existing_source_widgets` since we are just adding new source widgets and none of them will be selected by default. Existing source widgets should be irrelevant since we are just adding new source widgets.

After upating the signature, you can add back all the changes to the `update` function and pass the list of sources where the uuid is not in the list of widget_uuids (just the sources that need to be added). This is one way to support both optimizations: (1) scheduling when there many source widgets to add, (2) no longer recreating existing source widgets every sync. Another way would be to only run this method on the first run of the client.

Since we are syncing up about client architecture today, I'll provide relevant notes if there are any.

Reply:
No relevant updates from our architecture discussion not already mentioned in the comment above.

Reply:
OK... I'm working on this. To be clear, here's what I'm doing:

* Simplifying the code I've already done (see function signature).
* Only call this on application start.
* Any further source updates are as they currently are.
* (Also, rebase). :-)

That's it.

Reply:
That sounds good to me, although technically it's possible to hit a scenario where the client needs to download a number of sources that blocks the UI (does this start at 1000?) after application start: (1) the client starts in offline mode, (2) the user logs in and the client begins downloading 1000 sources, or (1) user logs in, (2) server is restarted because of a technical issue causing the client to be unable to download sources, (3) once server restarts the client starts downloading 1000 sources.  

If it's not too much work, we could cover more situations if you call your source add function when we add new sources from a sync.

Reply:
Looks like you addressed my comment by running the initial update if there are no sources in the source list, which covers both the offline case and the error-during-source-download case. I'll mark this as resolved.